python/adjoint/wrapper.py

```diff
@@ -156,13 +156,8 @@ def _run_fwd_simulation(self, design_variables):
         self.simulation.run(**sim_run_args)
 
         monitor_values = utils.gather_monitor_values(self.monitors)
-        fwd_fields = utils.gather_design_region_fields(
-            self.simulation,
-            design_region_monitors,
-            self.frequencies,
-        )
         return (jnp.asarray(monitor_values),
-                jax.tree_map(jnp.asarray, fwd_fields))
+                fwd_design_region_monitors)
```

Comment:
@ianwilliamson as you can see, we no longer pull the fields onto each process. Rather, we just hold onto opaque pointers which are later passed to `_get_gradient` (`fwd_design_region_monitors` and `adj_design_region_monitors` are the pointers) .

Do I need to do anything special to ensure jax plays nicely with this opaque data structures (before, you were using `tree_map`s and some other rather interesting data structures to wrap the arrays)?

Reply:
I think I got it, thanks.

Reply:
Sorry, I somehow missed this comment. If the the JAX test passes then I think it should be okay. The typical approach is to explicitly register the object type (or some class wrapping the object)) as a pytree.

Reply:
No worries. It works now, thanks. 

The key here is to keep it within the class object itself since we want it to remain opaque to jax.